### DISCUSSION TITLE
How do I deal with parallel fetch calls populating a select dropdown?

### DISCUSSION THREAD

**Discussion by svsaraf** [alpinejs/alpine#4144] [1. Help] [answered]
Hi folks! I'm a bit of a novice Alpine programmer so forgive me for the naive question. I have a select box which is structured like so:
```
<div x-data="{
        engagement: {},
        clients: [],
        init() {
           //each variable is populated with a fetch. So two fetches, in this case. For now displaying these as static.
            this.clients = [
                    {'name': 'John', 'pid': '2303'},
                    {'name': 'Jill', 'pid': '4032'},
                    {'name': 'Jane', 'pid': '6343'}
                ];
           this.engagement = {
                   'id': '345', 'client': '4032', 'engagementname': 'random name'
            }
        }
    }">
<select id="client" name="client" x-model="engagement.client"
            x-bind:disabled="clients.length === 0 || Object.keys(engagement).length === 0"
            class="mt-2 block w-full rounded-md border-0 py-1.5 pl-3 pr-10 text-gray-900 ring-1 ring-inset ring-gray-300 focus:ring-2 focus:ring-indigo-600 sm:text-sm sm:leading-6">
            <template x-for="client in clients" :key="client.pid">
                <option :value="client.pid" x-text="client.name"></option>
            </template>
</select>
```

Here's the problem. Depending on which fetch call returns first, the select dropdown is populated with the wrong data (or rather, the alpine variable engagement.client is **correct** but the wrong name is displayed in the dropdown. I've tried a bunch of tricks, including checking to make sure the fetches returned before showing the select (using a signal variable), disabling the dropdown if clients and engagement are empty (as above), but I can't guarantee the order in which the fetches return and no solution seems to work. If the clients fetch returns _second_, then "John" populates the dropdown, instead of "Jill". 

Any ideas on how you would solve? I've also replicated the issue on the page here: https://app.withpartly.com/labs/selectbug/

**Comment by svsaraf** [comment]
I eventually solved this problem by nesting the fetch calls together (i.e. call the second fetch after the first fetch returns). But that makes it difficult for me to "componentize" the various models. Ideally, the clients array lives on the select, while the engagement lives on the parent div. So the ideal solution would allow me to do independent fetch calls, while waiting for both to complete before hydrating the component.

**Comment by ekwoka** [comment, marked as answer]
Just use Promise.all to await the promises at the same time.

So you can have them in parallel and avoid race conditions.

**Reply by svsaraf** [reply]
Thanks!
